arangod/Metrics/ClusterMetricsFeature.cpp

```diff
@@ -0,0 +1,160 @@
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Valery Mironov
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Metrics/ClusterMetricsFeature.h"
+#include "Cluster/ClusterFeature.h"
+#include "Cluster/ClusterMethods.h"
+#include "Metrics/MetricsFeature.h"
+#include "Network/NetworkFeature.h"
+#include "Scheduler/SchedulerFeature.h"
+
+namespace arangodb::metrics {
+namespace {
+
+using namespace std::chrono_literals;
+constexpr std::chrono::steady_clock::duration kTimeout = 15s;
+
+}  // namespace
+
+ClusterMetricsFeature::ClusterMetricsFeature(
+    application_features::ApplicationServer& server)
+    : ApplicationFeature{server, "ClusterMetrics"} {
+  setOptional();
+  startsAfter<ClusterFeature>();
+  startsAfter<NetworkFeature>();
+  startsAfter<SchedulerFeature>();
+}
+
+void ClusterMetricsFeature::validateOptions(
+    std::shared_ptr<options::ProgramOptions> /*options*/) {
+  if (!ServerState::instance()->isCoordinator()) {
+    disable();
+  }
+}
+
+void ClusterMetricsFeature::asyncUpdate() {
+  if (isEnabled() && _count.fetch_add(1, std::memory_order_acq_rel) == 0) {
+    forceAsyncUpdate();
+  }
+}
+
+void ClusterMetricsFeature::forceAsyncUpdate() {
+  auto internal = [this] {
+    _count.store(1, std::memory_order_relaxed);
+    // TODO try get from agency
+    metricsOnCoordinator(server().getFeature<NetworkFeature>(),
+                         server().getFeature<ClusterFeature>())
+        .then([this](futures::Try<RawDBServers>&& metrics) mutable {
+          bool needRetry = !metrics.hasValue();
+          if (!needRetry) {
+            _lastUpdate = std::chrono::steady_clock::now();
+            // We want more time than kTimeout should expire after last cross
+            // cluster communication
+            set(std::move(*metrics));
+            needRetry = _count.fetch_sub(1, std::memory_order_acq_rel) != 1;
+            // if true then was call asyncUpdate() after _count.store(1)
+          }
+          if (needRetry) {
+            forceAsyncUpdate();
+          }
+        });  // detach() here
+  };
+  auto now = std::chrono::steady_clock::now();
+  if (now < _lastUpdate + kTimeout) {
```

Comment:
As far as I can see, `forceAsyncUpdate()` is private, but it can be called at any time by the public method `asyncUpdate()`. So it seems that there can be concurrent calls to `forceAsyncUpdate()`.
If that's true, then `_lastUpdate` and `_handle` should be protected by some synchronization, because they are otherwise not thread-safe.

Reply:
forceAsyncUpdate cannot running parallel